Features/plugin energy estimation tdp (#62)
alumet-dev/alumet: plugin-energy-estimation-tdp/README.md

Author: I suggest to change the plugin name:
EnergyEstimationTdp instead of EnergyEstimationTdpPlugin

Author: yes it is more coherent with other plugins

> # energy estimation tdp plugin

Reviewer: ```suggestion
# Energy estimation TDP plugin <!-- omit in toc -->
```

Reviewer: Toc is for table of content

Reviewer: This flag prevent the vscode extension we are using to add these titles in the table of content

## Table of Contents <!-- omit in toc -->

> - [introduction](#Introduction)
> - [Energy estimation tdp plugin](#Energy-estimation-tdp-plugin)
>   - [How to use](#how-to-use)
>   - [Prepare your environment](#prepare-your-environment)
>   - [Configuration](#Configuration)

Reviewer: ```suggestion
- [Introduction](#introduction)
- [Energy estimation tdp plugin](#energy-estimation-tdp-plugin)
  - [How to use](#how-to-use)
  - [Prepare your environment](#prepare-your-environment)
  - [Configuration](#configuration)
```

## Introduction

This plugin estimate the energy consumption for pods based on the TDP value of the machine where pods are running.

> The TDP (Thermal Design Power) is the maximum amount of heat generated by a computer compnent. 

Reviewer: ```suggestion
The TDP (Thermal Design Power) is the maximum amount of heat generated by a computer component.
```

Author: ok

Refer to https://en.wikipedia.org/wiki/Thermal_design_power for more details.

In the first version of the plugin, we consider only the TDP of the CPU.

>  This plugin requires the cgroupv2 input plugin (k8s) as it needs the input measurements of cgroupsv2.

Reviewer: ```suggestion
This plugin requires the cgroup v2 input plugin (k8s) as it needs the input measurements of cgroups v2.
```

Author: ok

>  The estimation calculation is done using the following formula:

Reviewer: ```suggestion
The estimation calculation is done using the following formula:
```

Author: ok

>  energy estimation = cgroupv2_cpu_total_usage * nb_vcpu/nb_cpu * tdp / 1000 000 / polling_interval

Reviewer: Can't we use latex to have a correctly displayed formula? Markdown support latex code blocks and/or simple formula

Author: Yes you are right , I found a solution but still some problems to display space, to be investigated.
Below the result:
![latex-md](https://github.com/user-attachments/assets/d752e9f2-9e77-41d8-b52c-0efcf6e09a60)

Reviewer 2: As Titouan said, [it should work right into the markdown document](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/writing-mathematical-expressions) :smiley: No need to use an external website (which would not work without internet, or after some time). Try to write the formula directly in the markdown source:
```md
$$latex here$$
```

cgroupv2_cpu_total_usage:   total usage of CPU in micro seconds for a pod 
nb_vcpu:                    number of virtual CPU of the hosting machine where pod is running
nb_cpu:                     number of physical CPU of the hosting machine where pod is running

> polling_interval:           polling interval of cgroupv2 input plugin

Reviewer: ```suggestion
polling_interval:           polling interval of cgroup v2 input plugin
```

Author: ok

> 

Reviewer: ```suggestion
```

Author: ok

## Energy estimation tdp plugin

### How to use

Just compile the app-agent of the alumet's github repository.

```bash
cargo run
```

> 

Reviewer: ```suggestion
```

Author: ok

The binary created by the compilation will be found under the target repository.

### Prepare your environment

To work this plugin needs k8s plugin configured, so the needed things are related to k8s plugin requirements:

1. cgroup v2
2. kubectl
3. alumet-reader user

> ### Configuration 

Reviewer: ```suggestion
### Configuration
```

Author: I do not understand how did you find the space at end line ? is there a trick ?

Reviewer: You need to have the following extensions installed in your vscode: 

[markdownlint](https://marketplace.visualstudio.com/items?itemName=DavidAnson.vscode-markdownlint)
and
[Markdown All in One](https://marketplace.visualstudio.com/items?itemName=yzhang.markdown-all-in-one)

[plugins.EnergyEstimationTdpPlugin]
enable = true
poll_interval = "30s"
tdp = 100.0
nb_vcpu = 1.0
nb_cpu = 1.0

> pool_interval: must be identical to the poll_interval of input k8s plugin. Default value is 

Reviewer: Yes? What's its default value?

nb_vcpu: number of virtual cpu allocated to the virtual machine in case of kubernetes nodes are virtual machine. Using the kubectl get node command, you can retrieve the number of vcpu. If the kubernetes nodes are physical machine, assign it to value 1. Default value is 1.

local@edge-master:~/manu/git/alumet$ kubectl  describe node edge-master.seed | grep cpu -B 2
  Hostname:    edge-master.seed
Capacity:

>   cpu:                32

Reviewer: What's this for?

Author: In fact it was to get the maximum CPU capacity allocated to a kubernetes node. I change it to be more generic.

nb_cpu: number of physical cpu of the hosted machine. Using the lscpu or hwinfo, you can retrieve the number of cpu. If the kubernetes nodes are physical machine, assign it to value 1. Default value is 1.

> tdp: Thermal Design power; each CPU has a calculated thermal design value; the value can be find on internet (usually on CPU manufacturer); you need the exact CPU family (using command lscpu or hwinfo). For example, for Intel® Xeon® D Processor, family D-2183IT, the tpd can be found:

Reviewer: hwinfo should be added to the project-words dictionary

https://ark.intel.com/content/www/us/en/ark/products/136441/intel-xeon-d-2183it-processor-22m-cache-2-20-ghz.html
tdp value is  100W.

> Default value is 100.

Reviewer: ```suggestion
Default value is 100.

```

Reviewer 2: I think he wants to add a trailing newline (standard)

Reviewer: Yes exactly
